src/helpers/privacy_budget.rs

```diff
@@ -0,0 +1,231 @@
+use redis::{transaction, Commands, RedisWrite, ToRedisArgs};
+use rust_elgamal::{DecryptionKey, RistrettoPoint, Scalar};
+#[derive(Debug, PartialEq, Eq)]
+struct BlindedMatchKey(Ciphertext);
+
+impl BlindedMatchKey {
+    pub fn new(key: DecryptionKey, message: RistrettoPoint, scalar: Scalar) -> Self {
+        let c = key.encryption_key().encrypt_with(message, scalar);
+        Self(c)
+    }
+}
+
+impl ToRedisArgs for BlindedMatchKey {
+    fn write_redis_args<W>(&self, out: &mut W)
+    where
+        W: ?Sized + RedisWrite,
+    {
+        let (c0, c1) = self.0.inner();
+        out.write_arg(&[*(c0.compress().as_bytes()), *(c1.compress().as_bytes())].concat());
+    }
+}
+
+/// This module implements privacy budget using a redis cache
+struct PrivacyBudget {
+    redis_connection: redis::Connection,
+    total_privacy_budget: u32,
+}
+
+impl PrivacyBudget {
+    fn new(redis_host_name: &str, redis_password: &str, is_tls: bool) -> Res<PrivacyBudget> {
+        //if Redis server needs secure connection
+        let uri_scheme = if is_tls { "rediss" } else { "redis" };
+
+        let redis_conn_url = format!("{}://:{}@{}", uri_scheme, redis_password, redis_host_name);
+
+        let redis_connection = redis::Client::open(redis_conn_url)?.get_connection()?;
+
+        Ok(PrivacyBudget {
+            redis_connection,
+            total_privacy_budget: 100,
+        })
+    }
+
+    /// Input : a vector of ciphertexts alongwith budget to be consumed.
+    /// This atomically increases the budget of all of the blinded match keys which have
+    /// sufficient budget to consume. For those which do not pass, match keys are discarded and not passed to the output.
+    pub fn get_eligible_packets_with_budget<'a>(
+        &'a mut self,
+        packets: impl Iterator<Item = &'a BlindedMatchKey>,
+        budget: u32,
+    ) -> impl Iterator<Item = &'a BlindedMatchKey> {
+        packets.filter(
+            move |packet| match self.increase_budget_consumed(packet, budget) {
+                Ok(true) => true,
+                Ok(false) | Err(_) => false,
+            },
+        )
+    }
+
+    pub fn get_total_privacy_budget(&self) -> u32 {
+        self.total_privacy_budget
+    }
+
+    /// Budget is increased for each key atomically in Redis
+    fn increase_budget_consumed(&mut self, packet: &BlindedMatchKey, budget: u32) -> Res<bool> {
+        let total_privacy_budget = self.get_total_privacy_budget();
+
+        Ok(transaction(
+            &mut self.redis_connection,
+            &[packet],
+            |con, pipe| {
+                let val: u32 = con.get(&packet).unwrap_or(0);
+                if val + budget > total_privacy_budget {
+                    pipe.set(&packet, val).ignore().query::<()>(con)?;
+                    Ok(Some(false))
+                } else {
+                    // increment
+                    pipe.set(&packet, val + budget).ignore().query::<()>(con)?;
+                    Ok(Some(true))
+                }
```

Comment:
This is not how I remember us discussing this working.

I remember us saying that we would start by just atomically incrementing all of the keys by the desired amount, then we would look at the return values to see which had exceeded the privacy budget, and subsequently decrementing those. That approach is more safe in a concurrent environment.

This approach isn't using atomic increments - it's first calling "con.get" to read the current value, then calling pipe.set(val + budget). That's not an atomic increment. If another query runs between those two lines, the value could be read at the pre-update value, then the new value overwritten, thereby allowing a query to exceed the privacy budget.

Reply:
Given that this is done in a transaction, the keys will be watched, increment will happen in memory and will be changed only if they are not changed by anyone else.
If the values are observed to be changed, redis will error out. 
This approach makes the entire operation concurrency safe as compared to increment and decrement

Reply:
That is so cool. Thanks for explaining this to me!